### ISSUE TITLE
vicadmin fails to launch with multiple datacenters in vCenter

### ISSUE THREAD

**Issue by malginin** [vmware/vic#2725]
**VIC Version:**

build 5760

**Code**

```ts
time="2016-10-13T21:12:03Z" level=debug msg="[BEGIN] [main.findDatastore:413]"
time="2016-10-13T21:12:03Z" level=debug msg="[BEGIN] [main.client:392]"
time="2016-10-13T21:12:03Z" level=warning msg="Unable to populate session: Failure finding dc (): default datacenter resolves to
multiple instances, please specify\nFailure finding ds (tintri1): please specify a datacenter\nFailure finding host (): please sp
ecify a datacenter"
time="2016-10-13T21:12:03Z" level=debug msg="[ END ] [main.client:392] [54.58396ms] "
time="2016-10-13T21:12:03Z" level=debug msg="[ END ] [main.findDatastore:413] [56.214483ms] "
time="2016-10-13T21:12:03Z" level=debug msg="[ END ] [main.configureReaders:101] [56.32614ms] "
time="2016-10-13T21:12:03Z" level=debug msg="[ END ] [main.(*server).serve:126] [56.729814ms] "
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x40c36c]

goroutine 1 [running]:
panic(0xc22040, 0xc420010060)
	/usr/local/go/src/runtime/panic.go:500 +0x1a1
main.findDatastore(0x0, 0x0)
	/drone/src/github.com/vmware/vic/cmd/vicadmin/vicadm.go:421 +0x13c
main.configureReaders(0x0)
	/drone/src/github.com/vmware/vic/cmd/vicadmin/logs.go:154 +0x1219
main.(*server).serve(0xc420269a40, 0x0, 0x0)
	/drone/src/github.com/vmware/vic/cmd/vicadmin/server.go:163 +0x85b
main.main()
	/drone/src/github.com/vmware/vic/cmd/vicadmin/vicadm.go:491 +0x3ab

goroutine 17 [syscall, locked to thread]:
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:2086 +0x1

goroutine 6 [syscall]:
os/signal.signal_recv(0xc42001bfa8)
	/usr/local/go/src/runtime/sigqueue.go:116 +0x157
os/signal.loop()
	/usr/local/go/src/os/signal/signal_unix.go:22 +0x22
created by os/signal.init.1
	/usr/local/go/src/os/signal/signal_unix.go:28 +0x41

goroutine 7 [IO wait]:
net.runtime_pollWait(0x7f5bf844a158, 0x72, 0x0)
	/usr/local/go/src/runtime/netpoll.go:160 +0x59
net.(*pollDesc).wait(0xc420214530, 0x72, 0xc42002fcf8, 0xc420010170)
	/usr/local/go/src/net/fd_poll_runtime.go:73 +0x38
net.(*pollDesc).waitRead(0xc420214530, 0x163a800, 0xc420010170)
	/usr/local/go/src/net/fd_poll_runtime.go:78 +0x34
net.(*netFD).accept(0xc4202144d0, 0x0, 0x1638e80, 0xc42021aaa0)
	/usr/local/go/src/net/fd_unix.go:419 +0x238
net.(*TCPListener).accept(0xc42001f158, 0xc420205a10, 0xc42002fe08, 0x916260)
	/usr/local/go/src/net/tcpsock_posix.go:132 +0x2e
net.(*TCPListener).AcceptTCP(0xc42001f158, 0xc42006f15c, 0xc42002fdf8, 0x607b60)
	/usr/local/go/src/net/tcpsock.go:209 +0x49
net/http.tcpKeepAliveListener.Accept(0xc42001f158, 0xc4202059e0, 0xc13880, 0x1626e50, 0xd16440)
	/usr/local/go/src/net/http/server.go:2608 +0x2f
net/http.(*Server).Serve(0xc42006f100, 0x163db40, 0xc42001f158, 0x0, 0x0)
	/usr/local/go/src/net/http/server.go:2273 +0x1ce
net/http.(*Server).ListenAndServe(0xc42006f100, 0xc42006f100, 0x0)
	/usr/local/go/src/net/http/server.go:2219 +0xb4
net/http.ListenAndServe(0xc42021a927, 0xe, 0x0, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/http/server.go:2351 +0xa0
github.com/vmware/vic/lib/pprof.StartPprof.func1(0xc42021a927, 0xe)
	/drone/src/github.com/vmware/vic/lib/pprof/pprof.go:95 +0x4b
created by github.com/vmware/vic/lib/pprof.StartPprof
	/drone/src/github.com/vmware/vic/lib/pprof/pprof.go:96 +0x25a

goroutine 8 [select, locked to thread]:
runtime.gopark(0xe53ac0, 0x0, 0xde48bc, 0x6, 0x18, 0x2)
	/usr/local/go/src/runtime/proc.go:259 +0x13a
runtime.selectgoImpl(0xc42001c730, 0x0, 0x18)
	/usr/local/go/src/runtime/select.go:423 +0x11d9
runtime.selectgo(0xc42001c730)
	/usr/local/go/src/runtime/select.go:238 +0x1c
runtime.ensureSigM.func1()
	/usr/local/go/src/runtime/signal1_unix.go:304 +0x2f3
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:2086 +0x1

goroutine 9 [chan receive]:
main.main.func1(0xc420051ec0, 0xc420269a40)
	/drone/src/github.com/vmware/vic/cmd/vicadmin/vicadm.go:486 +0x57
created by main.main
	/drone/src/github.com/vmware/vic/cmd/vicadmin/vicadm.go:489 +0x39a

goroutine 18 [IO wait]:
net.runtime_pollWait(0x7f5bf8449f18, 0x72, 0x6)
	/usr/local/go/src/runtime/netpoll.go:160 +0x59
net.(*pollDesc).wait(0xc42028d1e0, 0x72, 0xc4200306f0, 0xc420010170)
	/usr/local/go/src/net/fd_poll_runtime.go:73 +0x38
net.(*pollDesc).waitRead(0xc42028d1e0, 0x163a800, 0xc420010170)
	/usr/local/go/src/net/fd_poll_runtime.go:78 +0x34
net.(*netFD).Read(0xc42028d180, 0xc4202b4000, 0x1000, 0x1000, 0x0, 0x163a800, 0xc420010170)
	/usr/local/go/src/net/fd_unix.go:243 +0x1a1
net.(*conn).Read(0xc420266268, 0xc4202b4000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/net.go:173 +0x70
crypto/tls.(*block).readFromUntil(0xc42028fa10, 0x7f5bf844a410, 0xc420266268, 0x5, 0xc420266268, 0x28)
	/usr/local/go/src/crypto/tls/conn.go:472 +0x91
crypto/tls.(*Conn).readRecord(0xc420272380, 0xe53b17, 0xc420272480, 0xc4200001a0)
	/usr/local/go/src/crypto/tls/conn.go:574 +0xc4
crypto/tls.(*Conn).Read(0xc420272380, 0xc4202b5000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/usr/local/go/src/crypto/tls/conn.go:1109 +0x116
net/http.(*persistConn).Read(0xc42000b000, 0xc4202b5000, 0x1000, 0x1000, 0x6083c0, 0xc420030b58, 0x4134ad)
	/usr/local/go/src/net/http/transport.go:1261 +0x154
bufio.(*Reader).fill(0xc420287080)
	/usr/local/go/src/bufio/bufio.go:97 +0x10c
bufio.(*Reader).Peek(0xc420287080, 0x1, 0xc420030bbd, 0x1, 0x0, 0xc4202870e0, 0x0)
	/usr/local/go/src/bufio/bufio.go:129 +0x62
net/http.(*persistConn).readLoop(0xc42000b000)
	/usr/local/go/src/net/http/transport.go:1418 +0x1a1
created by net/http.(*Transport).dialConn
	/usr/local/go/src/net/http/transport.go:1062 +0x4e9

goroutine 19 [select]:
net/http.(*persistConn).writeLoop(0xc42000b000)
	/usr/local/go/src/net/http/transport.go:1646 +0x3bd
created by net/http.(*Transport).dialConn
	/usr/local/go/src/net/http/transport.go:1063 +0x50e

goroutine 26 [runnable]:
net.runtime_pollWait(0x7f5bf8449e58, 0x72, 0x8)
	/usr/local/go/src/runtime/netpoll.go:160 +0x59
net.(*pollDesc).wait(0xc4202d3480, 0x72, 0xc4200316f0, 0xc420010170)
	/usr/local/go/src/net/fd_poll_runtime.go:73 +0x38
net.(*pollDesc).waitRead(0xc4202d3480, 0x163a800, 0xc420010170)
	/usr/local/go/src/net/fd_poll_runtime.go:78 +0x34
net.(*netFD).Read(0xc4202d3420, 0xc420301000, 0x1000, 0x1000, 0x0, 0x163a800, 0xc420010170)
	/usr/local/go/src/net/fd_unix.go:243 +0x1a1
net.(*conn).Read(0xc4202662b8, 0xc420301000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/net.go:173 +0x70
crypto/tls.(*block).readFromUntil(0xc4202bddd0, 0x7f5bf844a410, 0xc4202662b8, 0x5, 0xc4202662b8, 0x28)
	/usr/local/go/src/crypto/tls/conn.go:472 +0x91
crypto/tls.(*Conn).readRecord(0xc420273180, 0xe53b17, 0xc420273280, 0xc4200001a0)
	/usr/local/go/src/crypto/tls/conn.go:574 +0xc4
crypto/tls.(*Conn).Read(0xc420273180, 0xc42030b000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/usr/local/go/src/crypto/tls/conn.go:1109 +0x116
net/http.(*persistConn).Read(0xc42000b400, 0xc42030b000, 0x1000, 0x1000, 0x6083c0, 0xc420031b58, 0x4134ad)
	/usr/local/go/src/net/http/transport.go:1261 +0x154
bufio.(*Reader).fill(0xc4202d5320)
	/usr/local/go/src/bufio/bufio.go:97 +0x10c
bufio.(*Reader).Peek(0xc4202d5320, 0x1, 0xc420031bbd, 0x1, 0x0, 0xc4203f2300, 0x0)
	/usr/local/go/src/bufio/bufio.go:129 +0x62
net/http.(*persistConn).readLoop(0xc42000b400)
	/usr/local/go/src/net/http/transport.go:1418 +0x1a1
created by net/http.(*Transport).dialConn
	/usr/local/go/src/net/http/transport.go:1062 +0x4e9

goroutine 27 [runnable]:
net/http.(*persistConn).writeLoop(0xc42000b400)
	/usr/local/go/src/net/http/transport.go:1646 +0x3bd
created by net/http.(*Transport).dialConn
	/usr/local/go/src/net/http/transport.go:1063 +0x50e
```

**Expected behavior:**

vicadmin starts and is accessible on tcp/2378

**Actual behavior:**

vic-init keeps relaunching vicadmin, vicadmin keeps crashing.

I noticed that vic-init doesn't pass "-dc" to vicadmin in the command line. Here's /vic/lib/install/management/appliance.go:

```
	conf.AddComponent("vicadmin", &executor.SessionConfig{
		User:  "vicadmin",
		Group: "vicadmin",
		Cmd: executor.Cmd{
			Path: "/sbin/vicadmin",
			Args: []string{
				"/sbin/vicadmin",
				"-docker-host=unix:///var/run/docker.sock",
				// FIXME: hack during config migration
				"-insecure",
				"-ds=" + conf.ImageStores[0].Host,
				"-cluster=" + settings.ClusterPath,
				"-pool=" + settings.ResourcePoolPath,
				"-vm-path=" + vm2.InventoryPath,
			},
			Env: []string{
				"PATH=/sbin:/bin",
				"GOTRACEBACK=all",
			},
			Dir: "/home/vicadmin",
		},
		Restart: true,
	},

```
On the other hand, session.finder.DatacenterOrDefault needs explicit dc in order not to fail in a multi-dc environment.

**Comment by mdubya66** [comment]
@sgairo can you take a look?

**Comment by sgairo** [comment]
Possibly related to #2729 , will take a closer look.

**Comment by dougm** [comment]
Not related to 2729, the problem is described in the summary: "-dc" flag not being set.  To reproduce, run the following against your VC: `govc datacenter.create foo` - any name will do, just need > 1 DC.

**Comment by chengwang86** [comment]
@dougm @malginin 
- I have one esx server and a vCenter on nimbus. 
- Then I created two DCs using "govc datacenter.create". 
- I added the esx server to the DC1 and configured everything for DC1. 
- For DC2, I didn't do any configuration; nor did it have any esx server.
- Then I ran "vic-machine-linux create" and it proceeded smoothly. 

Did I missing anything or is this problem not always reproducible?
